Openless PSA crypto APIs implementation

ronald-cron-arm opened the pull request:
## Description
This PR is the second part of the work related to #3265, the first part being #3527.

This PR implements the openless PSA crypo APIs that take as input or return a key identifier instead of a key handle (see #3265 for details).

## Status
Done

## Requires Backporting
No, PSA only

ronald-cron-arm commented:
The CI is as good as it can. Apart from Mbed OS tests, "merge TLS Testing" and "pr-merge" are falling because of the "ABI-API" testing as expected as this PR changes the definition of API parameter types.

gilles-peskine-arm reviewed (changes requested):
The basic design is sound, but there are major issues where the code as it currently is would be unsuitable for production.

Given the size of the changes, I strongly prefer additional commits to rebasing. I'll allow rewriting commits if you insist, but they would make the re-review a lot more difficult.

ronald-cron-arm commented:
>Given the size of the changes, I strongly prefer additional commits to rebasing. I'll allow rewriting commits if you insist, but they would make the re-review a lot more difficult.

Just a word on this. I am on the side of the reviewers on this. As a reviewer you prefer additional commits, I'll do additional commits. As the author of the PR, I have a deep enough understanding of the changes to be comfortable with both approaches. As long as there is not "too much" back and forth (I don't know if this is going to be the case or not) it should be fine for the second reviewer with additional commits as well. If it is not the case and the second reviewer would be more comfortable with a rewrite of the history, I may do one at some point with the goal of the result of the rewriting to be the same as what you last reviewed.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -88,10 +88,11 @@ psa_status_t psa_initialize_key_slots( void )
-    for( key = 1; key <= PSA_KEY_SLOT_COUNT; key++ )
+    size_t slot_idx;
+
+    for( slot_idx = 1; slot_idx <= PSA_KEY_SLOT_COUNT; slot_idx++ )
```
There is one backward loop over slots in `psa_get_empty_key_slot()`. I've kept this one for PSA_KEY_SLOT_COUNT to 1 as it is easier for the stop condition given that the loop index is an unsigned integer.

ronald-cron-arm replied:
Addressed in commit; "slot mgmt: Count from 0 when looping through slots"

ronald-cron-arm commented on include/psa/crypto_values.h:
```diff
@@ -1732,8 +1743,45 @@ static inline int mbedtls_svc_key_id_equal( mbedtls_svc_key_id_t id1,
+
 #endif /* !MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER */
 
+#define PSA_KEY_HANDLE_INIT MBEDTLS_SVC_KEY_ID_INIT
```
> I don't see the point of PSA_KEY_HANDLE_INIT and psa_key_handle_is_null. Using 0 for invalid/unset is a widespread convention and doesn't need named constants.
It is because a handle can be of integer or structure type thus 0 does not work. On that note, if before this openless work, the handles had been initialized with a PSA_KEY_HANDLE_INIT like macro it would have eased this openless work (as I wouldn't have had to introduce it). Thus, "using 0 for invalid/unset is a widespread convention and doesn't need named constants", is maybe not a very good idea if you are not completely sure that the integer value will always be an integer value.

>If they're defined at all, they should not be in crypto_values.h and they should hav an MBEDTLS_ prefix, since they aren't part of the standard API. They should be defined in library/something.h since they are neither part of the standard API, nor a useful API extension (they're new, but key handles are a legacy feature thta is only preserved for backward compatibility), nor a preserved legacy compatibility feature.

I am going to move them to crypto_compat.h along with the definition of `psa_key_handle_t` as they are tightly coupled to the `psa_key_handle_t` type. Because of this tight coupling, I find it strange to change there name. Thus, not doing it yet, but can do if you still prefer after this move to crypto_compat.h.

ronald-cron-arm replied:
psa_key_handle_equal has been removed.

ronald-cron-arm replied:
Addressed in commit: "Move handle related definition to crypto_compat.h"

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -51,31 +51,37 @@ typedef struct
-    *p_slot = slot;
-    return( PSA_SUCCESS );
+#if defined(MBEDTLS_PSA_CRYPTO_STORAGE_C)
+static psa_key_slot_t* psa_get_slot_from_key_id(
```
Indeed, there was room for improvement. I merged `psa_get_slot_from_transient_key_id()` and `psa_get_slot_from_key_id()` and named the function `psa_search_slot`.

ronald-cron-arm replied:
Addressed in commit; "Merge search of loaded transient and persistent keys"

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1777,11 +1789,12 @@ static psa_status_t psa_validate_key_attributes(
     psa_key_creation_method_t method,
     const psa_key_attributes_t *attributes,
-    psa_key_handle_t *handle,
+    mbedtls_svc_key_id_t *key,
```
I agree it is redundant but without it we would have to repeat the assignation of the key identifier from the slot to key identifier variables in callers: `my_key_identifier_variable = slot->attr.id`. Thus I've kept it and added a comment in the function description emphasizing the duplication.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -51,32 +51,37 @@ typedef struct
-    /* If the slot isn't occupied, the handle is invalid. */
-    if( ! psa_is_key_slot_occupied( slot ) )
-        return( PSA_ERROR_INVALID_HANDLE );
+    return( mbedtls_svc_key_id_equal( key, slot->attr.id ) ? slot : NULL );
```
This code is part of `psa_search_key_in_slot()` now and there is a comment in the function description about the special case of transient key identifier. This makes things about this code clearewr I think thus not changing anything yet but feel free to ask again for it.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -51,32 +51,37 @@ typedef struct
-    /* If the slot isn't occupied, the handle is invalid. */
-    if( ! psa_is_key_slot_occupied( slot ) )
-        return( PSA_ERROR_INVALID_HANDLE );
+    return( mbedtls_svc_key_id_equal( key, slot->attr.id ) ? slot : NULL );
```
Indeed. I've added tests checking that:
. the key owner of an imported volatile key is the one specified when importing a key.
. a key identifier of a volatile key with a valid PSA key identifier but the wrong owner is rejected.

ronald-cron-arm replied:
Addressed in commit: "Add tests checking owner of volatile keys"

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1242,10 +1242,10 @@ psa_status_t psa_destroy_key( psa_key_handle_t handle )
         return( PSA_SUCCESS );
 
-    status = psa_get_key_slot( handle, &slot );
+    status = psa_get_key_slot( key, &slot );
```
We need to load it to know if it is a SE key or not, right?

gilles-peskine-arm replied:
Ah, good point. Please add a comment.

ronald-cron-arm replied:
Comment added in commit: "psa: Do not reset a key slot under access"

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -700,14 +700,8 @@ void invalid_handle( int handle_construction,
-                invalid_handle = 2;
-            else
-                invalid_handle = valid_handle - 1;
+            invalid_handle =
```
This was not an unopened invalid handle. I moved back to something close to what is in development currently and added a comment.

ronald-cron-arm replied:
Addressed in commit: "tests: slot mgmt: Fix unopened handle build"

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_slot_management.data:
```diff
@@ -13,54 +13,15 @@ transient_slot_lifecycle:0:0:PSA_KEY_TYPE_RAW_DATA:"0123456789abcdef0123456789ab
 Transient slot, check after restart with live handles
 transient_slot_lifecycle:0:0:PSA_KEY_TYPE_RAW_DATA:"0123456789abcdef0123456789abcdef":CLOSE_BY_SHUTDOWN
 
```
I've restored all the test cases. We still support open/close thus this clean-up was indeed premature. I've also restored some code in `persistent_slot_lifecycle`. The code is eventually not the same as in development (usage of key identifier instead of key handle essentially) but the checks are the same but when checking that an handle is not valid anymore after a close because this is not true anymore.

ronald-cron-arm replied:
Addressed in commits:
Revert "tests: Remove some invalid persistent key handle tests"
tests: slot mgmt: Restore close related code

ronald-cron-arm commented on library/psa_crypto_slot_management.h:
```diff
@@ -22,27 +22,44 @@
- * \param handle        Key handle to query.
+ *  The last PSA_KEY_SLOT_COUNT identifiers of the implementation range
+ *  of key identifiers are reserved for transient key identifiers.
+ *  A transient key identifier is equal to PSA_KEY_ID_TRANSIENT_MIN plus the
```
No, I don't think there is a significant benefit. Not change for the time being but if you prefer I can change to not using transient.

ronald-cron-arm commented on include/psa/crypto.h:
```diff
@@ -526,12 +529,12 @@ psa_status_t psa_close_key(psa_key_handle_t handle);
- * \param[out] target_handle On success, a handle to the newly created key.
- *                          \c 0 on failure.
+ * \param[out] target_key   On success, an identifier for the newly created
+ *                          key. PSA_KEY_ID_NULL on failure.
```
I moved back to `\c 0`.

ronald-cron-arm replied:
Addressed in commit: "Remove usage of PSA_KEY_ID_NULL for zero key identifiers"

ronald-cron-arm commented on library/pk.c:
```diff
@@ -150,11 +150,12 @@ int mbedtls_pk_setup( mbedtls_pk_context *ctx, const mbedtls_pk_info_t *info )
     const mbedtls_pk_info_t * const info = &mbedtls_pk_opaque_info;
     psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
-    psa_key_handle_t *pk_ctx;
+    mbedtls_svc_key_id_t *pk_ctx;
```
Are you sure for (cipher¦pk)*.[hc] ? As PSA code is for example calling mbedtls_cipher_free() this is not really PSA client code, right?

gilles-peskine-arm replied:
The cipher and pk modules currently contain both client-side code and server-side code because parts of them are above the PSA crypto API layer and parts are underneath. As part of the psa-isaztion of the library, all remaining calls to cipher and pk from `psa_crypto.c` should disappear eventually, and then cipher and pk will be client-side only. An opaque cipher or pk object is purely client-side.

ronald-cron-arm replied:
If MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER is set then the first argument of psa_get_key_attributes() (called by mbedtls_pk_setup_opaque()) is of type mbedtls_svc_key_id_t (!= psa_key_id_t) thus the 2nd parameter of mbedtls_pk_setup_opaque() as to be of type mbedtls_svc_key_id_t and pk_ctx as well. Thus if MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER  is set, I cannot move here to psa_key_id_t.

But, maybe it is just the case that mbedtls_pk_setup_opaque() should not be defined if MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER  is defined, thus incompatibility between  MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER  and MBEDTLS_USE_PSA_CRYPTO ?

gilles-peskine-arm replied:
The cipher/pk opaque interface is intended as a transitional interface to facilitate the migration of existing code (such as the X.509 and TLS parts of Mbed TLS) to PSA. In a world where crypto is in its own service, it wouldn't need to exist on the server side. So indeed I don't have a use case for `MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER` and `MBEDTLS_USE_PSA_CRYPTO` together.

ronald-cron-arm replied:
Thus I guess that the way forward is for MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER not to be set in the full configuration as incompatible with MBEDTLS_USE_PSA_CRYPTO and add in all.sh a test component dedicated to MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER: full - MBEDTLS_USE_PSA_CRYPTO  + MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER where only the unit tests are run. Does it sound good?

gilles-peskine-arm replied:
Actually [`MBEDTLS_USE_PSA_CRYPTO` should not be enabled in `full`](https://github.com/ARMmbed/mbedtls/issues/3603), so `MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER` could be. But in the short term, I guess it's better to keep `MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER` out of `full`, with a note that it should be added back once `MBEDTLS_USE_PSA_CRYPTO` is removed.

ronald-cron-arm replied:
I've removed MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER from the full configuration and added a component in all.sh to test the configuration (full - MBEDTLS_USE_PSA_CRYPTO + MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER).

ronald-cron-arm replied:
Addressed in commit:
lib: Remove usage of mbedtls_svc_key_id_t
State PSA_CRYPTO_KEY_ID_ENCODES_OWNER and USE_PSA_CRYPTO incompatibility

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -244,67 +245,21 @@ void persistent_slot_lifecycle( int lifetime_arg, int owner_id_arg, int id_arg,
     TEST_EQUAL( psa_get_key_attributes( handle, &read_attributes ),
                 PSA_ERROR_DOES_NOT_EXIST );
-    psa_reset_key_attributes( &read_attributes );
     TEST_EQUAL( psa_close_key( handle ), PSA_ERROR_DOES_NOT_EXIST );
```
Restored.

ronald-cron-arm replied:
Addressed in commit: "tests: slot mgmt: Restore close related code"

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1718,17 +1718,28 @@ static psa_status_t psa_validate_key_attributes(
     if( status != PSA_SUCCESS )
         return( status );
 
+    /* Validate the key identifier only in the case of a persistent key. */
```
This is mitigated by the fact that psa_set_key_id() forces the key to be permanent and psa_set_key_lifetime() forces the key identifier to be 0 if the key is volatile: thus difficult to end up with a non zero key identifier in attributes to create a volatile key. Anyway when copying attributes from a created volatile key (happens in the tests) it is still possible to end up with a non zero key identifier in attributes to create a volatile key thus I've added the check. As a consequence I had to fix some tests and even library code that was not handling the location part of lifetimes properly (like psa_set_key_id() ...).

ronald-cron-arm replied:
Addressed in commits:
Add key id check when creating a volatile key
psa: Fix tests/handling of lifetime incompatible with location

ronald-cron-arm commented on programs/psa/crypto_examples.c:
```diff
@@ -92,7 +92,7 @@ static psa_status_t cipher_operation( psa_cipher_operation_t *operation,
 }
 
-static psa_status_t cipher_encrypt( psa_key_handle_t key_handle,
+static psa_status_t cipher_encrypt( mbedtls_svc_key_id_t key,
```
If I do what you suggest here, it means that PSA programs don't compile if MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER is set. In those programs we are testing our PSA crypto implementation whose key identifier type is `mbedtls_svc_key_id_t` not `psa_key_id_t`. That's the way it is thus I am not sure why this is a problem in fact.

ronald-cron-arm replied:
In fact it is more exact to say that programs test/demonstrate our implementation of the PSA API specification and thus that key identifier should be of type psa_key_id_t. Thus move from mbedtls_svc_key_id_t to psa_key_id_t as suggested and don't compile programs using PSA APIs when MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER is set.

ronald-cron-arm replied:
Addressed in commit: "program: Remove usage of mbedtls_svc_key_id_t"

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1718,17 +1718,28 @@ static psa_status_t psa_validate_key_attributes(
+    {
+        status = psa_validate_key_id(
+            psa_get_key_id( attributes ),
+            psa_key_lifetime_is_external( lifetime ) );
```
I've changed it to not possible to create a key with an id in the vendor range even with `mbedtls_psa_register_key_id`, updated se tests accordingly and added tests to check that the vendor range is not accessible for persistent keys in local storage.

ronald-cron-arm replied:
Addressed in commit: "psa: Forbid creation/registration of keys in vendor range"

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1142,7 +1142,7 @@ static psa_status_t psa_get_key_from_slot( psa_key_handle_t handle,
 
-    status = psa_get_key_slot( handle, &slot );
+    status = psa_get_key_slot( key, &slot );
     if( status != PSA_SUCCESS )
```
I'm not sure what you mean here. If the key identifier is not correct, there is nothing to remove. In case of failure for internal reasons (insufficient memory for example) maybe but I would say such error handling shouldn't be here.

gilles-peskine-arm replied:
Ok: `psa_get_key_slot` cleans up after itself — it won't return with a partially allocated slot. So there's no need for any cleanup here in `psa_get_key_from_slot`.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -88,26 +94,34 @@ psa_status_t psa_initialize_key_slots( void )
+                                ( (psa_key_id_t)slot_idx ) - 1;
+
             return( PSA_SUCCESS );
+        }
```
I started with a simple solution where a key slot containing a permanent key description could be reused at any time as a single-threading only solution. But it didn't work with the copy of a persistent key where the key slot containing the description of the persistent key to copy can not be reused while the copy is on-going. Thus I changed gear and went for counting current accesses to a key slot. This is a bit invasive and overkill for just the single-threaded case probably but as we will need it for the multi-threaded case ...

ronald-cron-arm replied:
Addressed in commits:
psa: mgmt: Add key slot reuse
psa: Do not reset a key slot under access
psa: Add access counter check in slot wipe
psa: slot mgmt: Add unaccessed slots counter in stats
psa: mgmt: Add key slot access counter

ronald-cron-arm commented on docs/getting_started.md:
```diff
@@ -76,7 +76,7 @@ void import_a_key(const uint8_t *key, size_t key_len)
     psa_status_t status;
     psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
-    psa_key_handle_t handle;
+    mbedtls_svc_key_id_t key;
```
Addressed in Commit "docs: Use psa_key_id_t instead of mbedtls_svc_key_id_t"

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -214,18 +188,29 @@ psa_status_t psa_validate_key_persistence( psa_key_lifetime_t lifetime,
     }
 }
 
+psa_status_t psa_validate_key_id( mbedtls_svc_key_id_t key, int vendor_ok )
```
Addressed in commit: "slot mgmt: Define psa_validate_key_id before it's used"

ronald-cron-arm commented on library/ssl_tls.c:
```diff
@@ -521,7 +520,7 @@ static int tls_prf_generic( mbedtls_md_type_t md_type,
      * and for this use case it makes sense to have a 0-length "secret".
      * Since the key API doesn't allow importing a key of length 0,
-     * keep master_slot=0, which setup_psa_key_derivation() understands
+     * keep master_key=NULL, which setup_psa_key_derivation() understands
```
Addressed in commit: "Remove usage of PSA_KEY_ID_NULL for zero key identifiers"

ronald-cron-arm commented on include/psa/crypto_types.h:
```diff
@@ -247,6 +247,12 @@ typedef struct
+ * To support temporary both openless APIs and psa_open_key(), define
+ * psa_key_handle_t to be equal to mbedtls_svc_key_id_t.
+ */
+typedef mbedtls_svc_key_id_t psa_key_handle_t;
```
Addressed in commit: "Move handle related definition to crypto_compat.h"

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -150,33 +164,50 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
+    if( *p_slot != NULL )
+        return( PSA_SUCCESS );
+
+#if defined(MBEDTLS_PSA_CRYPTO_STORAGE_C)
```
Addressed in commit: "slot mgmt: Update psa_get_key_slot documentation"

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -264,7 +296,7 @@ psa_status_t psa_close_key( psa_key_handle_t handle )
+    if( psa_key_handle_is_null( handle ) )
         return( PSA_SUCCESS );
 
     status = psa_get_key_slot( handle, &slot );
```
Addressed in commit: "Fix close/purge of a key"

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -245,7 +246,13 @@ int check_key_attributes_sanity( psa_key_handle_t key )
 
     /* Persistence */
     if( lifetime == PSA_KEY_LIFETIME_VOLATILE )
```
Addressed in commit: "tests: psa: Use PSA_KEY_LIFETIME_IS_VOLATILE where it should"

ronald-cron-arm commented on include/psa/crypto.h:
```diff
@@ -646,8 +644,8 @@ psa_status_t psa_destroy_key(psa_key_handle_t handle);
  *                          it must be equal to the size from \p data.
- * \param[out] handle       On success, a handle to the newly created key.
- *                          \c 0 on failure.
+ * \param[out] key          On success, an identifier to the newly created key.
```
Addressed in commit: "psa: Improve key creation documentation"

ronald-cron-arm commented on library/psa_crypto_slot_management.h:
```diff
@@ -46,21 +46,20 @@
 #define PSA_KEY_ID_TRANSIENT_MAX  PSA_KEY_ID_VENDOR_MAX
 
-/** Access a key slot at the given handle.
+/** Access a key slot for a given key identifier.
```
Addressed in commit: "slot mgmt: Update psa_get_key_slot documentation"

ronald-cron-arm commented on programs/psa/key_ladder_demo.c:
```diff
@@ -541,42 +539,42 @@ static psa_status_t run( enum program_mode mode,
-    /* Close any remaining key. Deinitializing the crypto library would do
-     * this anyway, but explicitly closing handles makes the code easier
+    /* Destroy any remaining key. Deinitializing the crypto library would do
+     * this anyway, but explicitly destroying leys makes the code easier
```
Addressed in commit: "Fix comment in key_ladder_demo.c"

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -6,11 +6,11 @@
+    CLOSE_BY_DESTROY, /**< Destroy key(s) */
+    CLOSE_BY_SHUTDOWN, /**< Deinit and reinit without closing keys. */
+    CLOSE_BY_CLOSE_WITH_SHUTDOWN, /**< Close key(s) then deinit/reinit. */
+    CLOSE_BY_DESTROY_WITH_SHUTDOWN, /**< Destroy key(s) then deinit/reinit. */
```
Addressed in commits:
tests: slot mgmt: Add psa_purge_key testing
tests: slot mgmt: Fix test data
tests: slot mgmt: Rename ways of invalidating keys
tests: slot mgmt: Add purge checks in volatile key lifecycle test

ronald-cron-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -534,8 +534,8 @@ psa_set_key_type(&attributes, PSA_KEY_TYPE_AES);
 psa_set_key_usage_flags(&attributes, PSA_KEY_USAGE_ENCRYPT | PSA_KEY_USAGE_DECRYPT);
-psa_key_handle_t handle = 0;
-psa_generate_key(&attributes, &handle);
+mbedtls_svc_key_id_t key;
```
Addressed in commit: "docs: Use psa_key_id_t instead of mbedtls_svc_key_id_t"

ronald-cron-arm commented:
I believe this is ready for re-review. I first pushed new patches addressing @gilles-peskine-arm  comments ending up at commit [13c5f76](https://github.com/ronald-cron-arm/mbedtls/tree/psa-openless-1). All the CI tests passed on this commit but there was conflicts in merging it into the `development` branch.

I then rebased on top of the `development` branch ending up at  [bd22ab7](https://github.com/ARMmbed/mbedtls/commit/bd22ab7ec30c1d1c8df27ef7360c59862efe1cae)  resolving the five conflicts with `development`. The CI ran successfully on that commit (both head and merge tests but for the ABI-API-checking test as expected).

I then force-pushed some changes in test code that was added and used handles instead of key identifiers. The diff of this last step can be seen [here](https://github.com/ARMmbed/mbedtls/compare/bd22ab7ec30c1d1c8df27ef7360c59862efe1cae..cd65c6cbcbac319b67923a0ca99c8da2d4cfbfb). On this commit all tests are passing but in the merge test job: ABI-API-checking test and three other tests because of CI infrastructure glitches. As those 3 last tests were passing on the penultimate version I don't feel the need to re-trigger the CI and think it is ok like this.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -877,3 +877,189 @@ exit:
+    i = PSA_KEY_SLOT_COUNT;
+    key = mbedtls_svc_key_id_make( i, i + 1 );
+    psa_set_key_id( &attributes, key );
+
+    if( lifetime == PSA_KEY_LIFETIME_VOLATILE )
+        psa_set_key_lifetime( &attributes, PSA_KEY_LIFETIME_VOLATILE );
```
Why not unconditionally
```
psa_set_key_lifetime( &attributes, lifetime );
```

ronald-cron-arm replied:
I think I did it like that because psa_set_key_id() set the lifetime to PERSISTENT thus the call to psa_set_key_lifetime() is not necessary in the case of a persistent key, but it doesn't hurt thus I removed the test.

gilles-peskine-arm replied:
That would make sense if this test was testing that the default lifetime is volatile, but since it isn't, I think it's better to simplify the code by removing the condition.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -877,3 +877,189 @@ exit:
+                                (uint8_t *) &i, sizeof( i ),
+                                &returned_key_id ) );
+    if( lifetime != PSA_KEY_LIFETIME_VOLATILE )
+        TEST_ASSERT( mbedtls_svc_key_id_equal( returned_key_id, key ) );
```
And if the lifetime is volatile, it would be good to check that the id is in the appropriate range.

gilles-peskine-arm reviewed (changes requested):
The overall design looks good and I'm glad you made significant headway on slot access management with the access count. I have several localized change requests but nothing major.

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1189,36 +1189,48 @@ static psa_status_t psa_restrict_key_policy(
+ * psa_decrement_key_slot_access_count() when it does not access the key slot
+ * anymore.
+ */
+static psa_status_t psa_get_key_from_slot( mbedtls_svc_key_id_t key,
```
For the names of the functions, I like the `acquire` verb. Otherwise:

. to replace psa_search_key_in_slots, not completely convinced by psa_acquire_key_in_memory. Other possibilities: psa_acquire_key_in_slot(s) or psa_acquire_loaded_key() ?
. to replace psa_get_key_slot maybe that psa_load_and_acquire_key() would be better than just psa_acquire_key() emphasizing the possible load. The weakness is that there may be no loading.
. an alternative to  psa_acquire_key_with_policy would then be psa_load_and_acquire_key_with_policy()

What do you think about those suggestions?

gilles-peskine-arm replied:
* “Acquire key in slots” could mean a key that's acquired from slots or that's placed into slots. “Acquire loaded key” isn't ambiguous, but it looks like it excludes volatile keys. Maybe `psa_acquire_key_from_slots`? But I still prefer “memory”, because storage could have slots as well.
* Indeed “load and acquire” is a little misleading since it won't load already-loaded persistent keys and volatile keys. I do like having some disambiguation on just ”acquire“, but I don't think “load and acquire” is right. `psa_acquire_key_with_load`?

ronald-cron-arm replied:
ok thus last but not least. Eventually, I am not sure about `acquire` because when you acquire something you own it but here it is not the case, you don't own the key slot, somebody else can get a read access to it.

First regarding psa_get_key_slot(), from the point of view of the user, the user has got a key identifier and she wants the description of the key corresponding to that identifier. As key descriptions are stored in what we call key slots in the implementation, psa_get_key_slot()  seems at least a good base for the name.

Then, still from the point of view of the user whether the function has to load data for persistent storage or not does not really matter thus better not to change/tweak the base name for that.

On the contrary, the function also increments the key slot access counter and this is important for the user to be aware of that as it is her responsibility to decrement the counter when she's done with the key description. The access counter is there to prevent the key slot to be wiped out and being used for another purpose while a part of the library is making read accesses to it. Thus as you mentioned in [this comment](https://github.com/ARMmbed/mbedtls/pull/3547#discussion_r523406314) from the user point of view it is about locking the key slot to prevent it from being wiped out/reused while using it. Thus what about psa_get_and_lock_key_slot() ?

Then for psa_search_key_in_slots(), it would be psa_get_and_lock_key_slot_in_memory() and for psa_get_get_from_slot, psa_get_and_lock_key_slot_with_policy().

Basically I am proposing to replace `acquire` by `get_and_lock` in the function names you proposed. Otherwise, we would have  `psa_decrement_key_slot_access_count` → `psa_lock_key_slot`, `psa_increment_key_slot_access_count` → `psa_unlock_key_slot`.

What do you think?

gilles-peskine-arm replied:
You can acquire a semaphore multiple times, so I don't think “acquire” is exclusive. But I'm ok with lock/unlock as well.

bensze01 replied:
Of the options discussed, I think I prefer the "aquire" / "release" verbs. I don't think that "load and aquire" is much more misleading than "aquire with load". I'd go with "load and aquire"  due to the interaction with the "with policy" name
As for the "from slots" / "from memory" question I have no strong feelings one way or the other. Maybe go with slots, since it's already used in the key slot functions?

So I'd probably go with:
psa_acquire_key_in_slots (/ psa_acquire_key_in_memory)
psa_load_and_acquire_key
psa_load_and_acquire_key_with_policy
psa_acquire_key_slot
psa_release_key_slot

ronald-cron-arm commented:
I've simplified the commit history (from 49 to 37 commits) and forced-pushed the result. The head is now [def441e](https://github.com/ARMmbed/mbedtls/commit/def441e7293ccc1edd59b2216c7222364ed8f0c9). It is equivalent to the previous head [cd65c6c](https://github.com/ARMmbed/mbedtls/commit/cd65c6cbcbac319b67923a0ca99c8da2d4cfbfb1) (the previous head can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-openless-2)) but for the renaming of "transient key identifiers" to "volatile key identifier" and the diff can be found [here](https://github.com/ARMmbed/mbedtls/compare)./cd65c6cbcbac319b67923a0ca99c8da2d4cfbfb1..def441e7293ccc1edd59b2216c7222364ed8f0c9).

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -73,44 +89,52 @@ static void psa_purge_key_storage( void )
-        case CLOSE_BY_DESTROY:
-        case CLOSE_BY_DESTROY_WITH_SHUTDOWN:
-            PSA_ASSERT( psa_destroy_key( handle ) );
+        /* Purging the key just purge RAM data of persitent keys. */
```
and also persitent → persistent

ronald-cron-arm replied:
Yes I spotted that one as well.

gilles-peskine-arm commented on include/psa/crypto_struct.h:
```diff
@@ -374,9 +374,17 @@ static inline struct psa_key_attributes_s psa_key_attributes_init( void )
 static inline void psa_set_key_id( psa_key_attributes_t *attributes,
```
Not worth a rebase on its own, but you left “not taking account” which should be “not taken into account”.

gilles-peskine-arm reviewed (commented):
I reviewed the rebase from cd65c6c to def441e and verified that it does [what Ronald claimed](https://github.com/ARMmbed/mbedtls/pull/3547#issuecomment-724811549).

ronald-cron-arm commented on library/psa_crypto_core.h:
```diff
@@ -36,6 +36,33 @@
+     * may access it. For example, such control is needed in the following
+     * scenarios:
+     * . In case of key slot starvation, all key slots contain the description
+     *   of a key, and the library asks for the description of a permanent
```
Indeed, I didn't realize that I was completely diverging from the current wording.

ronald-cron-arm commented:
New force-push from from def441e to a5b894f to fix the commit message of commit "psa: Fix tests/handling of lifetime incompatible with location". No code change thus the force-push diff is empty.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -51,30 +51,101 @@ typedef struct
 
-    if( ! global_data.key_slots_initialized )
-        return( PSA_ERROR_BAD_STATE );
+    psa_status_t status = psa_validate_key_id( key, 1, 1 );
```
There is no `else` branch in `psa_key_id_is_volatile` thus I am not sure what you mean here.

gilles-peskine-arm replied:
Sorry, I meant the else branch of `psa_search_key_in_slots`. Change
```
psa_validate_key_id()
if psa_key_id_is_transient()
  <check slot[id]>
else
  <search slots for id>
```
to
```
if psa_key_id_is_transient()
  <check slot[id]>
else
  psa_validate_key_id()
  <search slots for id>
```

ronald-cron-arm replied:
Thanks for the clarification and yes this is a good idea.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -270,21 +271,68 @@ void persistent_slot_lifecycle( int lifetime_arg, int owner_id_arg, int id_arg,
+    }
 
 exit:
     PSA_DONE( );
```
The call here returns in error and I can see the reset being done in `psa_get_key_attributes` when the function returns in error thus it seems we are ok here without the reset.

ronald-cron-arm replied:
But there is another one missing thus adding attribute resets in clean-up part.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -51,30 +51,101 @@ typedef struct
+    if( transient_ok &&
+        ( PSA_KEY_ID_TRANSIENT_MIN <= key_id ) &&
+        ( key_id <= PSA_KEY_ID_TRANSIENT_MAX ) )
+        return( PSA_SUCCESS );
```
Code removed.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -112,7 +112,7 @@ int lifetime_is_secure_element( psa_key_lifetime_t lifetime )
     /* At the moment, anything that isn't a built-in lifetime is either
      * a secure element or unassigned. */
-    return( lifetime != PSA_KEY_LIFETIME_VOLATILE &&
+    return( ( ! PSA_KEY_LIFETIME_IS_VOLATILE( lifetime ) ) &&
```
I can fix it but even in the full configuration (`MBEDTLS_PSA_CRYPTO_SE_C` enabled) `lifetime_is_secure_element` always returns `false`. I would be on the side of removing it but I don't know the whole story about it, let me know.

gilles-peskine-arm replied:
It turns out that `test_suite_psa_crypto.data` doesn't have any test case with a key in a secure element. But if it did, `lifetime_is_[dynamic_]secure_element` would serve its purpose, which is (so far) to identify keys that have a slot number attribute.

Please leave this code around since it's incomplete but useful.

ronald-cron-arm replied:
ok fair enough, fixed and changed the name to `lifetime_is_dynamic_secure_element`.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -257,6 +272,36 @@ psa_status_t psa_get_key_slot( mbedtls_svc_key_id_t key,
+    MBEDTLS_PARAM_FAILED( slot->access_count > 0 );
+#endif
+#ifdef MBEDTLS_PLATFORM_C
+    mbedtls_fprintf( stderr,
```
It seems that if MBEDTLS_PLATFORM_C is defined then mbedtls_fprintf() is defined. It may not do anything but it is define thus having this seems acceptable to me. As I explain in the comment above, the idea is to maximize our chance to spot when something goes wrong. In the comment I should probably say "try to output an error message on standard error output if possible" rather than just "output an error message on standard error output" though.
Regarding consistency the case of `psa_xxx_abort` functions is maybe a bit different as we don't expect their returned code to be ignored in some cases while we expect it for `psa_decrement_key_slot_access_count()`.

gilles-peskine-arm replied:
But outputting to stderr might not be right. For example, this might be a log file which must be in a certain format, or a log file that's visible to an attacker that we don't want to give information to. Adding stderr feedback is a functional and security design change.

ronald-cron-arm replied:
I am not completely convinced but I removed it as this discussion goes significantly beyond the scope of this PR I'd say.

ronald-cron-arm commented on library/psa_crypto_core.h:
```diff
@@ -36,6 +36,33 @@
+     *   or purge or destroy a key while it is in used by the library through
+     *   another thread.
+     */
+    size_t access_count;
```
I think eventually it is better to check for the overflow. It adds a bit a code for not much but that way we are such that an count overflow won't occur.

ronald-cron-arm commented on include/psa/crypto_extra.h:
```diff
@@ -231,6 +231,8 @@ typedef struct mbedtls_psa_stats_s
     /** Number of slots that are not used for anything. */
     size_t empty_slots;
+    /** Number of slots that are not accessed. */
+    size_t unaccessed_slots;
```
That way we check non-zero values which gives more confidence that the counting actually occurred. But I don't have a strong opinion about it and I can change it to `accessed_slots` if you prefer.

gilles-peskine-arm replied:
I don't understand this argument. Seeing a nonzero value doesn't tell us that “counting actually occurred”. If fields aren't getting marked as accessed, it'll get stuck at the total number of slots if it's counting unaccessed slots, and at zero if it's counting accessed slots.

gilles-peskine-arm replied:
On a related note, nooking at this with a fresh mind, I don't like “accessed”, regardless of which way it's counting. To me “access” evokes a read operation, but this is counting slots that are “locked” in some way. Following https://github.com/ARMmbed/mbedtls/pull/3547#discussion_r517638903 I suggest `acquired_slots` (or `unacquired_slots`, but I still don't see a reason for counting down).

ronald-cron-arm replied:
I am feeling a difference between the two but cannot put that rapidly into a logical explanation (indication that this is wrong?) and I don't think it is worth it. I said I can change it. Do you prefer me to change it to `accessed/acquired_slots` ?

gilles-peskine-arm replied:
Unless you can find an explicit argument for counting un-whatever slots, please change to `acquired_slots`.

ronald-cron-arm commented:
I've addressed all comments and the CI is passing but the ABI-API check as expected. Ready for review.

gilles-peskine-arm commented on ChangeLog.d/psa-openless.txt:
```diff
@@ -0,0 +1,19 @@
+   * The build system now checks that MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER
+     and MBEDTLS_USE_PSA_CRYPTO are not enabled simultaneously as those two
+     configuration options are incompatible.
+   * psa_set_key_id() no longer resets the key location when enforcing the
+     key lifetime to be PSA_KEY_LIFETIME_PERSISTENT.
```
That's not exactly it: it also doesn't enforce the persistent level to be `PERSISTENT`, just non-volatile.

> psa_set_key_id() no longer resets the lifetime of volatile keys in secure elements.

ronald-cron-arm replied:
I understand your point but not your suggestion. I've tried to improve.

gilles-peskine-arm commented on ChangeLog.d/psa-openless.txt:
```diff
@@ -0,0 +1,19 @@
+   * Attempting to create a volatile key with a non-zero key identifier now
+     fails. Previously the key identifier was just ignored when creating a
+     volatile key.
+   * Attempting to register a key in a secure element or to create a persistent
+     key with a key identifier in the vendor range now fails.
```
This looks like it's saying “Attempting to register a key in a secure element now fails”. Suggestion:

> Attempting to create or register a key with a key identifier in the vendor range now fails.

gilles-peskine-arm reviewed (changes requested):
I'm almost happy. Most of my remaining comments are about documentation.

I still don't like the terminology around the access counter/lock. Please wait for a third opinion before making more changes. @bensze01 What do you think about https://github.com/ARMmbed/mbedtls/pull/3547#discussion_r517638903 ?

ronald-cron-arm commented on tests/suites/test_suite_pk.function:
```diff
@@ -119,6 +119,7 @@ mbedtls_svc_key_id_t pk_psa_genkey( void )
     PSA_ASSERT( psa_generate_key( &attributes, &key ) );
 
 exit:
+    psa_reset_key_attributes( &attributes );
```
> But the interface permits not calling `psa_reset_key_attributes` if you've only used `psa_set_key_xxx` (excluding domain parameters) and not `psa_get_key_attributes`. So we should test that there's no memory leak in such cases.

I know that and I think I tried to explain in the commit message why any way I prefer to do it systematically for educational purposes. With this approach, I think that the probability that a user not reading the interface documentation (what !) does not call `psa_reset_key_attributes` when he must is smaller. The probability that he does call it when it is not necessary is greater but has it doesn't hurt.

gilles-peskine-arm replied:
My point is that if we systematically call `psa_reset_key_attributes` in tests, but there's a bug in the library that causes one of the functions like `psa_set_key_type` to allocate resources, we won't catch it.

ronald-cron-arm replied:
Ah, right. Then what about to not reset in tests focusing on `psa_set_xxx` APIs testing ?

gilles-peskine-arm replied:
Yes, at least those. But more generally I think reset() shouldn't be called in functions that don't set domain parameters and don't call get_attributes().

ronald-cron-arm replied:
I cannot say I don't understand this point of view but the fact is that even in the library tests thus in code that is written by people knowing reasonably well PSA APIs some necessary calls are missing. Thus, this was an attempt to do something about it. I think that any alloc/free API couple is quite systematically a source of lots of bugs and here I would say it is even worst as from the name `psa_get_key_attributes` it is not completely obvious that this function does some allocations. Thus still in the spirit of trying to improve things but being less invasive, I could limit what I have done to the test functions where it is necessary to call `psa_reset_key_attributes` and also add some comment with it. That way the reset call  would gain some visibility. What do you think?

gilles-peskine-arm replied:
We could instrument `psa_get_key_attributes` to allocate domain parameters under `MBEDTLS_TEST_HOOKS`. This way Asan would complain if we forgot reset(). But this is all out of scope here.

For this PR, I'd prefer to only add `psa_reset_key_attributes` in new cases and optionally in cases where it was necessary but forgotten before, but not to add it in places where it was definitely optional, and not to do any other related work.

ronald-cron-arm replied:
ok I wonder how to do that. I revert the commit adding the reset and add on top a new commit adding just a few resets. Then at the end when the PR is approved I rewrite the history removing the "reset commit" and its revert commit with an identical end result? Another possibility is to get rid of the reset commit now through a force-push but it may be disturbing for the end of the review. Let me know what you prefer.

gilles-peskine-arm replied:
I think it would be best to wait until Bence has finished reviewing, then modify the commit.

ronald-cron-arm commented on library/psa_crypto_slot_management.h:
```diff
@@ -22,32 +22,86 @@
+ * In case of a persistent key, the function loads the description of the key
+ * into a key slot if not already done.
+ *
+ * On success, the returned key slot is locked. It is the responsibility of
+ * the caller to unlock the key slot when it does not access it anymore.
```
As I said I am not convinced by `acquire`. You mentioned semaphore terminology but as you outline here a semaphore is to count down a limited amount of resources. This is not the case here. In its current form, the lock/counter is there only to prevent a key slot to be wiped out and reused while some read accesses are on-going. It counts the number of on-going read accesses to the key slot and thus when the counter is not equal to zero the key slot is locked in the sense that it cannot be wiped out and reused for another persistent key.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -88,29 +150,80 @@ psa_status_t psa_initialize_key_slots( void )
+    /*
+     * If there is no unused key slot and there is at least one unlocked key
+     * slot containing the description of a permament key, recycle the first
+     * such key slot we encountered. If we need later on to operate on the
```
ok thanks for pointing that to me, I will read about that in my English grammar to get a full picture.

bensze01 reviewed (changes requested):
The following section can be removed from `crypto.h`
```
/** \brief Key handle.
 *
 * This type represents open handles to keys. It must be an unsigned integral
 * type. The choice of type is implementation-dependent.
 *
 * 0 is not a valid key handle. How other handle values are assigned is
 * implementation-dependent.
 */
typedef _unsigned_integral_type_ psa_key_handle_t;
```

ronald-cron-arm commented on ChangeLog.d/psa-openless.txt:
```diff
@@ -0,0 +1,19 @@
+Features
+   * Use and access persistent keys directly from their key identifiers. As an
+     upgrade of the MbedTLS PSA Cryptography implementation from version 1.0
+     beta to the released version 1.0.0 it is now not necessary to open a
+     persistent key to use it. Opening a persistent key is still supported and
+     even not yet deprecated but will be deprecated in the coming months.
```
Just changed psa_key_identifier_t  to psa_key_id_t.

ronald-cron-arm commented:
This comment is indeed not at the right place (psa_key_handle_t is not platform specific now) and wrong as psa_key_handle_t is not an integer type anymore when MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER is defined. As this type is on its way out (now defined in crypto_compat.h) I think it is ok if it does not appear anymore in the DOXYGEN documentation. Thus I just removed this comment,

ronald-cron-arm commented:
I've force pushed a change to the last commit adding the change logs and added a new commit fixing and improving  documentation. Thus I believe I've addressed all comments but the ones related to access counter naming and key attributes reset. @bensze01 waiting on your final review and your opinion about the naming [discussion](https://github.com/ARMmbed/mbedtls/pull/3547#discussion_r523437919) to proceed.

gilles-peskine-arm commented on ChangeLog.d/psa-openless.txt:
```diff
@@ -0,0 +1,17 @@
+     releases.
+
+Bugfix
+   * psa_set_key_id() no longer resets the key location when enforcing the
+     key lifetime to be non-volatile.
```
We both got this wrong. `psa_set_key_id` didn't reset the key location (except during an intermediate version of this PR IIRC). The actual change is from
```
if persistence == VOLATILE and location == LOCAL: persistence = PERSISTENT
```
to
```
if persistence == VOLATILE: persistence = PERSISTENT
```

So a correct description would be

> psa_set_key_id() now also sets the lifetime to persistent for keys located in a secure element.

ronald-cron-arm replied:
Yes after looking at the version of the function in `development` I agree with that. I seems I got eventually lost in the various versions of the function.

ronald-cron-arm replied:
I've fixed that in the "Add change log" commit and force-pushed the result.

gilles-peskine-arm reviewed (changes requested):
I'm happy apart from one changelog entry and the lock/acquire terminology issue.

gilles-peskine-arm reviewed (dismissed):
Looks good to me apart from the lock/acquire terminology issue and https://github.com/ARMmbed/mbedtls/pull/3547#discussion_r525430529.

~I'm approving the PR~ I'm willing to approve the PR if the `psa_reset_key_attributes` testing issue is fixed because this is internal terminology and we can improve it in a follow-up.

ronald-cron-arm commented:
Forced push from  2a61649 to eb9a307:
2a61649 can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-openless-3).
Starting from 2a61649 I have removed the commit "tests: psa: Reset systematically key attributes" (873e688) and then added the new commit that add the psa_reset_key_attributes() where needed.

gilles-peskine-arm reviewed (dismissed):
Approved. I'd prefer to improve the lock/acquire terminology, but since this is purely internal, I'm ok with doing that in a follow-up.

ronald-cron-arm commented:
> Approved. I'd prefer to improve the lock/acquire terminology, but since this is purely internal, I'm ok with doing that in a follow-up.

Great. I agree with you that the lock/acquire terminology can and must be improved and I am going to work on it in the coming days. @bensze01 ok to approve on your side as well?

bensze01 commented on library/psa_crypto_core.h:
```diff
@@ -101,18 +100,17 @@ static inline int psa_is_key_slot_occupied( const psa_key_slot_t *slot )
  *
- * A key slot is accessed iff its access counter is strickly greater than
- * 0.
+ * A key slot is locked iff its lock counter is strickly greater than 0.
```
```suggestion
 * A key slot is locked iff its lock counter is strictly greater than 0.
```

bensze01 reviewed (approved):
I've left a couple of suggestions, but they are all just typo fixes in comments.
Sine no actual code changes are necessary, I'm approving this PR.

ronald-cron-arm commented:
Force push  from eb9a307 to 3a4f0e3:
I've just rework the penultimate commit "Improve/fix documentation" to address Bence's last comments.
You can find the diff between eb9a307 to 3a4f0e3 [here](https://github.com/ARMmbed/mbedtls/compare/eb9a307650ef5860b3b99c64ebf415178abb735f..3a4f0e3cc4bb8a289b724f415c10dd83cd2b9ffe).

gilles-peskine-arm reviewed (approved):
As before, approved on the condition of a follow-up that improves acquire/lock terminology.

bensze01 reviewed (approved):
All of my comments were addressed, so I'm approving this PR as well.

gilles-peskine-arm commented:
CI passed except for expected ABI/API changes (`psa_key_handle_t` → `psa_key_identifier_t` or `mbedtls_svc_key_id_t`, changes in internal structures). I filed https://github.com/ARMmbed/mbedtls/issues/3904 for the improvements around acquire/lock naming. This is ready for merge.
